### PR TITLE
quickly added form body support

### PR DESCRIPTION
**Pull request by stopthatastronaut**
As before, might need some tidying up - my C# chops aren't what they once were

One use case I had was to handle Slack outgoing webhooks. Unfortunately, they POST as x-www-form-urlencoded, not JSON.

So I quickly added some code for that

### CODE REVIEW

**Comment by tintoy** [comment]
That's great - I think we can probably detect `x-www-form-urlencoded` by the `Content-Type` header BTW.

**Comment by tintoy** [comment]
Give me a couple of minutes and I'll pull this in.

**Comment by stopthatastronaut** [comment]
Yeah, header would be more reliable I reckon. It _was_ a quick & dirty change ;)

File: Controllers/GenericController.cs [hunk 1]
```diff
@@ -209,10 +209,10 @@ public dynamic GetJob(string jobId)
                         }
                     }
 
-					if (method.Parameters.Any(param => queryStrings.All(q => q.Key != param.Name)))
+					if (method.Parameters.Any(param => queryStrings.All(q => q.Key != param.Name))) 
 					{
 						DynamicPowershellApiEvents.Raise.VerboseMessaging(String.Format("Cannot find all parameters required."));
-						throw new MissingParametersException("Cannot find all parameters required.");
```
Comment (tintoy) [inline Controllers/GenericController.cs:215]: Maybe check if the missing parameter is optional (`!Required`) and only throw if it's not?

Comment (tintoy) [reply Controllers/GenericController.cs:215]: If we allow required parameters to go through to Powershell without values then the call is going to fail anyway...

**Review by tintoy — CHANGES_REQUESTED** [review]
Need to think about the required-parameters change, it has the potential to break existing clients...

**Comment by tintoy** [comment]
Ok @stopthatastronaut - I've pushed some minor changes to enable optional properties (in the config file, mark the parameter with `IsOptional="true"`.

Can you have a look and let me know if this works for you?

**Comment by tintoy** [comment]
You can basically just tack on

```csharp
else if (requestContentType == "x-www-form-urlencoded")
{
	NameValueCollection formFields = await Request.Content.ReadAsFormDataAsync();
	foreach (string formFieldName in formFields.AllKeys)
	{
		commandParameters.Add(
			key: formFieldName,
			value: formFields[formFieldName]
		);
	}
}
```

and then this looks good to merge.

**Comment by tintoy** [comment]
Do you have time to resolve the merge conflict, @stopthatastronaut? If not, let me know and I'll manually alter the PR to fix it.

**Comment by tintoy** [comment]
Never mind - I manually merged it. Thanks for your contribution! 😄

**Pull request merged**